File: ip2cidr.py
# http://www.ipaddressguide.com/cidr

import math
def ipToVal(ip):
    ip = ip.split(".")
    val = 0
    for x in ip:
        val = (val << 8) + int(x)
    return val

def ValToIp(val):
    ip, i = ["0"] * 4, 3
    while val:
        ip[i] = str(val % (1<<8))
        val /= (1<<8)
        i -= 1
    return ".".join(ip)

def range2cidr(start, end):
    if not start  or not end or start.count('.') != 3 or end.count('.') != 3:
        return None
    start, end = ipToVal(start), ipToVal(end)
    if start > end:
        return None
    ans = []
    while start <= end:
        firstOne = start & (-start)
        maxMask = 32 - int(math.log(firstOne, 2))
        maxDiff = 31 - int(math.log(end - start+1 , 2))
        maxMask = max(maxMask, maxDiff)
        ip = ValToIp(start)
        ans.append(ip + "/" + str(maxMask))
        start += 2 ** (32 - maxMask)
    return ans

if __name__ == '__main__':
    start = '10.0.1.24'
    end = '10.0.1.225'
    a = range2cidr(start, end)
    print a

File: ip2cidr.cpp
long ipToVal(string ip) {
    long val = 0;
    int i = 0;
    for(int j=0; i<4 && j<ip.length; ++i){
        // 查找某一给定位置后的子串的位置
        auto nx = ip.find('.', j);
        if(nx == ip.npos){
            val = (val << 8) + atoi(ip.substr(j).c_str());
            ++i;
            break;
        }
        val = (val<<8) + atoi(ip.substr(j, nx - j).c_str());
        j = nx +1;
    
    }
    if(i!=4) throw "the ip is incorrect";
    return val;
}

string valToIp(long val){
    string ip = "";
    for(int i =0; i<4; ++i){
        ip = to_string(val % 256) + "." + ip;
        val /= 256;
    }
    ip.pop_back();
    return ip;
}

vector<string> range2cidr(string start, string end){
    long st = ipToVal(start), ed = ipToVal(end);
    vecto<string> ans;
    while(st <= ed) {
        int lastOne = st & (-st);
        int maxMask - 32 - int((log(startOne)/log(2)));
        int maxDiff = 32 - int(floor(log(ed-st+1)/log(2)));
        maxMAsk = max(maxMask, maxDiff);
        string ip = valToIp(st);
        ans.push_back(ip+'/'+to_string(maxMask));
        st += int(pow(2, 32-maxMask));
    }
    return ans;
}
